encoder/generator.go

```diff
@@ -476,68 +485,82 @@ func (g *generator) genInterface(name string, v cue.Value) {
 
 		kv := KV{K: k, V: vstr}
 
-		d, ok := fields.Value().Default()
-		// [...number] results in [], which is not desired
-		// TODO: There must be a better way to handle this
-
-		// Correct the list default value when it is not for real
-		if ok && d.Kind() == cue.ListKind {
-			len, err := d.Len().Int64()
-			if err != nil {
-				g.addErr(err)
-			}
-			var defaultExist bool
-			if len <= 0 {
-				op, vals := fields.Value().Expr()
-				if op == cue.OrOp {
-					for _, val := range vals {
-						vallen, _ := d.Len().Int64()
-						if val.Kind() == cue.ListKind && vallen <= 0 {
-							defaultExist = true
-							break
-						}
-					}
-					if !defaultExist {
-						ok = false
-					}
-				} else {
-					ok = false
-				}
-			}
-		}
+		exists, defaultV, err := tsPrintDefault(fields.Value())
+		g.addErr(err)
 
-		if ok {
-			dStr, err := tsprintField(d)
-			g.addErr(err)
-			kv.Default = dStr
-			if isReference(d) {
-				kv.Default = strcase.ToLowerCamel(kv.Default + "Default")
-			}
+		if exists {
 			tvars["defaults"] = true
+			kv.Default = defaultV
 		}
-
 		pairs = append(pairs, kv)
 	}
 
 	sort.Slice(pairs, func(i, j int) bool { return pairs[i].K < pairs[j].K })
 	tvars["pairs"] = pairs
-
 	g.exec(interfaceCode, tvars)
 }
 
+func tsPrintDefault(v cue.Value) (bool, string, error) {
+	var result string
+	d, ok := v.Default()
+	// [...number] results in [], which is a fake default, we need to correct it here.
+	if ok && d.Kind() == cue.ListKind {
+		len, err := d.Len().Int64()
+		if err != nil {
+			return false, "", err
+		}
+		var defaultExist bool
+		if len <= 0 {
+			op, vals := v.Expr()
+			if op == cue.OrOp {
+				for _, val := range vals {
+					vallen, _ := d.Len().Int64()
+					if val.Kind() == cue.ListKind && vallen <= 0 {
+						defaultExist = true
+						break
+					}
+				}
+				if !defaultExist {
+					ok = false
+				}
+			} else {
+				ok = false
+			}
+		}
+	}
+
+	if ok {
+		dStr, err := tsprintField(d)
+		if err != nil {
+			return false, result, err
+		}
+		result = dStr
+		if isReference(d) {
+			result = strcase.ToLowerCamel(result + "Default")
+		}
+		return true, result, nil
+	}
+	return false, result, nil
+}
+
 // Render a string containing a Typescript semantic equivalent to the provided
 // Value, if possible.
 //
 // The provided Value must be a simple expression (loosely defined, until
 // something more precise is understood); e.g., this will NOT render a struct
 // literal.
-func tsprintField(v cue.Value) (string, error) {
+func tsprintField(v cue.Value, optionals ...int) (string, error) {
```

Comment:
Why does this need to be variadic? Unless i'm missing something, it seems like it should be sufficient to just make it a single int corresponding to the nesting level.

Reply:
yes, just want to do something equivalent to default parameter to avoid update everywhere :D, apparently in go there is no such thing. Will do something static.